File: ninloader/python/ninloader/core/discovery.py
"""Discovery engine — scans all registered collectors for available token sources."""

from __future__ import annotations

import json
from typing import Dict, List, Optional

from ..types import DiscoveredToken
from ..collectors import CollectorRegistry


class DiscoveryEngine:
    """Iterates all registered collectors and runs discover() on each."""

    def run(
        self,
        service_filter: Optional[str] = None,
    ) -> List[DiscoveredToken]:
        """Run discovery across all (or filtered) collectors.

        Returns a list of DiscoveredToken for every token source found.
        """
        results: List[DiscoveredToken] = []

        for key, collector_cls in CollectorRegistry.get_all().items():
            svc, src = key
            if service_filter and svc != service_filter:
                continue

            collector = collector_cls()

            if not collector.is_platform_supported():
                continue

            try:
                discovered = collector.discover()
                results.extend(discovered)
            except Exception:
                # discover() should never raise, but just in case
                pass

        return results

    @staticmethod
    def format_table(tokens: List[DiscoveredToken]) -> str:
        """Format discovered tokens as a human-readable table."""
        if not tokens:
            return "No token sources discovered."

        lines = []
        lines.append(f"{'Service':<12} {'Source':<20} {'Stealth':<8} {'Account':<30} {'Path'}")
        lines.append("-" * 100)

        for t in tokens:
            acct = t.account_hint or ""
            path = t.path or ""
            details = f" ({t.details})" if t.details else ""
            lines.append(
                f"{t.service:<12} {t.source:<20} {t.stealth_score:<8} {acct:<30} {path}{details}"
            )

        lines.append(f"\nTotal: {len(tokens)} token source(s) found")
        return "\n".join(lines)

    @staticmethod
    def format_json(tokens: List[DiscoveredToken]) -> str:
        """Format discovered tokens as machine-readable JSON."""
        data = []
        for t in tokens:
            data.append({
                "service": t.service,
                "source": t.source,
                "path": t.path,
                "account_hint": t.account_hint,
                "stealth_score": t.stealth_score,
                "details": t.details,
            })
        return json.dumps(data, indent=2)

File: ninloader/go/internal/discovery/discovery.go
// Package discovery scans all registered collectors for available token sources.
package discovery

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

// Engine iterates all registered collectors and runs Discover() on each.
type Engine struct{}

// Run executes discovery across all (or filtered) collectors.
// Returns a list of DiscoveredToken for every token source found.
// If serviceFilter is non-empty, only collectors matching that service are run.
func (e *Engine) Run(serviceFilter string) []*types.DiscoveredToken {
	var results []*types.DiscoveredToken

	var collectors []interface {
		Service() string
		Source() string
		IsPlatformSupported() bool
		Discover() []*types.DiscoveredToken
	}

	if serviceFilter != "" {
		for _, c := range registry.AllForService(serviceFilter) {
			collectors = append(collectors, c)
		}
	} else {
		for _, c := range registry.All() {
			collectors = append(collectors, c)
		}
	}

	for _, c := range collectors {
		if !c.IsPlatformSupported() {
			continue
		}

		discovered := safeDiscover(c)
		results = append(results, discovered...)
	}

	return results
}

// safeDiscover calls Discover() on a collector, recovering from panics.
func safeDiscover(c interface {
	Discover() []*types.DiscoveredToken
}) (results []*types.DiscoveredToken) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Fprintf(os.Stderr, "[WARN] collector panic in discover: %v\n", r)
			results = nil
		}
	}()
	return c.Discover()
}

// FormatTable produces a human-readable table with columns:
// Service, Source, Stealth, Account, Path (Details).
func (e *Engine) FormatTable(tokens []*types.DiscoveredToken) string {
	if len(tokens) == 0 {
		return "No token sources discovered."
	}

	var b strings.Builder
	fmt.Fprintf(&b, "%-12s %-20s %-8s %-30s %s\n", "Service", "Source", "Stealth", "Account", "Path")
	b.WriteString(strings.Repeat("-", 100))
	b.WriteByte('\n')

	for _, t := range tokens {
		acct := t.AccountHint
		path := t.Path
		details := ""
		if t.Details != "" {
			details = fmt.Sprintf(" (%s)", t.Details)
		}
		fmt.Fprintf(&b, "%-12s %-20s %-8d %-30s %s%s\n", t.Service, t.Source, t.StealthScore, acct, path, details)
	}

	fmt.Fprintf(&b, "\nTotal: %d token source(s) found", len(tokens))
	return b.String()
}

// FormatJSON serializes the discovered tokens as a JSON array.
func (e *Engine) FormatJSON(tokens []*types.DiscoveredToken) string {
	data := make([]map[string]any, len(tokens))
	for i, t := range tokens {
		data[i] = map[string]any{
			"service":       t.Service,
			"source":        t.Source,
			"path":          t.Path,
			"account_hint":  t.AccountHint,
			"stealth_score": t.StealthScore,
			"details":       t.Details,
		}
	}

	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "[]"
	}
	return string(b)
}
